### PR TITLE
py/parsenum: support full complex parsing

### PR DESCRIPTION
**Pull request by dpgeorge**
This is an alternative to #7623, taking work done by @jimmo and extending it.

I improved @jimmo attempt further reducing it down to +40 bytes on stm32 (from his original WIP of +64; and #7623 is +96).

And then I also needed to add a second commit to fix parsing of `complex("nanj")` and `complex("infj")`, which were previously unsupported by would have parsed as `nan+j` without the second commit here.

The second commit also fixes `complex("j")` which is currently broken on master and returns `0j`.

Code size diff for first commit is:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:   +80 +0.015% 
unix nanbox:   +96 +0.021% 
      stm32:   +40 +0.010% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:    +0 +0.000% GENERIC
      esp32:   +36 +0.002% GENERIC
     mimxrt:   +40 +0.011% TEENSY40
 renesas-ra:   +40 +0.006% RA6M2_EK
        nrf:   +32 +0.017% pca10040
        rp2:   +32 +0.006% PICO
       samd:   +16 +0.011% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

Code size diff for second commit is:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:    +0 +0.000% 
unix nanbox:   +32 +0.007% 
      stm32:   +36 +0.009% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:   +24 +0.003% GENERIC
      esp32:   +36 +0.002% GENERIC
     mimxrt:   +32 +0.009% TEENSY40
 renesas-ra:   +32 +0.005% RA6M2_EK
        nrf:   +32 +0.017% pca10040
        rp2:   +32 +0.006% PICO
       samd:   +20 +0.014% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

In total this PR is (sum of above):
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:   +80 +0.015% 
unix nanbox:  +128 +0.028% 
      stm32:   +76 +0.019% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:   +24 +0.003% GENERIC
      esp32:   +72 +0.005% GENERIC
     mimxrt:   +72 +0.020% TEENSY40
 renesas-ra:   +72 +0.011% RA6M2_EK
        nrf:   +64 +0.034% pca10040
        rp2:   +64 +0.013% PICO
       samd:   +36 +0.026% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

Is it really worth it?

### CODE REVIEW

File: tests/float/complex1.py [hunk 1]
```diff
@@ -72,6 +80,13 @@
 print(float("inf") * (1 + 1j))
 print(float("-inf") * (1 + 1j))
 
+# malformed complex strings
+for test in ("1+2", "1j+2", "1+2j+3", "1+2+3j"):
```
Comment (dlech) [inline tests/float/complex1.py:84]: We could throw in `"1 + 2j"` for good measure.

File: tests/float/complex1.py [hunk 2]
```diff
@@ -4,9 +4,17 @@
 print(complex(1))
 print(complex(1.2))
 print(complex(1.2j))
+print(complex("j"))
```
Comment (dlech) [inline tests/float/complex1.py:7]: Should we add cases with upper-case J as well?

**Comment by jimmo** [comment]
This still feels like an exceptionally rarely-used feature... but compatibility is important, and fixing the `complex("j")` case is worth doing.

I think if we don't merge this then we should consider removing support for `complex(str)` altogether.

File: tests/float/complex1.py (hunk 1 above)
Comment (dpgeorge) [reply tests/float/complex1.py]: added

File: tests/float/complex1.py (hunk 2 above)
Comment (dpgeorge) [reply tests/float/complex1.py:7]: added

**Comment by dpgeorge** [comment]
I pushed another commit to reduce the size on builds that have complex disabled.  That commit has the following diff (ports not listed have 0 change):
```
   bare-arm:    -4 -0.007% 
minimal x86:    -2 -0.001% 
    esp8266:   -32 -0.005% GENERIC
       samd:   -60 -0.043% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

So now the overall change of this PR is:
```
   bare-arm:    -4 -0.007% 
minimal x86:    -2 -0.001% 
   unix x64:   +80 +0.015% 
unix nanbox:  +128 +0.028% 
      stm32:   +76 +0.019% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:    -8 -0.001% GENERIC
      esp32:   +72 +0.005% GENERIC
     mimxrt:   +72 +0.020% TEENSY40
 renesas-ra:   +72 +0.011% RA6M2_EK
        nrf:   +64 +0.034% pca10040
        rp2:   +64 +0.013% PICO
       samd:   -24 -0.017% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```

**Pull request merged**